scheduler/system_util.go

```diff
@@ -152,11 +144,12 @@ func diffSystemAllocsForNode(
 		// If we are on a tainted node, we must migrate if we are a service or
 		// if the batch allocation did not finish
 		if nodeIsTainted {
-			// If the job is batch and finished successfully, the fact that the
-			// node is tainted does not mean it should be migrated or marked as
-			// lost as the work was already successfully finished. However for
-			// service/system jobs, tasks should never complete. The check of
-			// batch type, defends against client bugs.
+			// If the job is batch and finished successfully (but not yet marked
+			// terminal), the fact that the node is tainted does not mean it
```

Comment:
Huh I'm surprised this is possible because of: https://github.com/hashicorp/nomad/blob/v1.5.0-beta.1/client/allocrunner/alloc_runner.go#L812-L843

When sending alloc updates from client->server, the client computes the alloc's ClientStatus based on the state of every task precisely so that from the scheduler's perspective `all tasks completed -> alloc completed` is in an atomic update and therefore an invariant.

This suggests that state transition isn't atomic/invariant... I don't mind being defensible against misbehaving clients here (as I think we should do that more!), but it does make me wonder if we should rethink the approach the linked code in the client does entirely.... perhaps the fsm should _also_ enforce the invariant that `all task states being terminal == alloc's client status is terminal`?

Reply:
It's entirely possible that it _isn't_ possible to reach this state. This block was introduced at the same time as the dead code I removed above that assumes we can have terminal allocations here at all, which we can't. I added the `(but not yet marked terminal)` parenthetical here because I assumed it was correct, but on reflection maybe it's unreachable outside of unit tests.

Reply:
Returning to this, I think we should probably address it, but I want to pull that out to a separate PR just because it's not _obviously_ correct to remove this and I don't want to muddy up this commit with that if we have to git bisect it later.